File: sdk/typescript/src/ctk/runner.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
/**
 * CTK runner: load vectors, drive a harness, assert `expect`.
 *
 * The assertion engine, capability skip check, and scripted
 * interceptor/resolver evaluation live in the Rust core (native.ctk*).
 * This module keeps only vector globbing, the recording wrapper, and
 * the orchestration loop that calls the native `Harness`.
 */

import { readFileSync, readdirSync } from "node:fs";
import { join } from "node:path";

import {
  AgentContext,
  ApprovalRequest,
  ApprovalResolution,
  Composition,
  CompositionConfig,
  EnforcementMode,
  Interceptor,
  JsonValue,
  Verdict,
} from "../index";
import { native } from "../native";
import type { Harness, RunRecord, Scenario } from "./index";

/** TODO(stage-4): vectors still authored in the pre-P-003 five-verdict
 * wire vocabulary (`warn`, `escalate`, `approval_resolver_missing`).
 * Stage 4 rewrites them to the three-verdict shapes (§5.1); until then
 * their scripted verdicts fail the §5 gate by design (fail closed) and
 * no longer exercise the seam/warning semantics they were written for,
 * so they are skipped here. */
const TODO_STAGE_4 = [
  "AH-CTK-030", // escalate-approve → deny+approval / resolution
  "AH-CTK-031", // escalate-reject → deny+approval / reject
  "AH-CTK-032", // escalate-no-resolver → liftable deny stands
  "AH-CTK-050", // warn-passthrough → allow+warnings
  "AH-CTK-072", // resolver-identity-mismatch → approval_identity_mismatch
  "AH-CTK-073", // resolver-raises → approval_resolver_failed
];

export interface VectorResult {
  id: string;
  title: string;
  status: "pass" | "fail" | "skip";
  detail: string;
  failures: string[];
}

export function loadVectors(dir: string): JsonValue[] {
  return readdirSync(dir)
    .filter((f) => /^AH-CTK-.*\.json$/.test(f))
    .sort()
    .map((f) => JSON.parse(readFileSync(join(dir, f), "utf8")) as JsonValue);
}

/** Replays one `interceptor_script` rule list via the Rust core. */
class ScriptedInterceptor implements Interceptor {
  protected readonly rulesJson: string;
  constructor(rules: JsonValue) {
    this.rulesJson = JSON.stringify(rules);
  }
  intercept(ctx: AgentContext): Verdict {
    const w = JSON.parse(native.ctkScriptedIntercept(this.rulesJson, JSON.stringify(ctx)));
    if (w !== null && typeof w === "object" && "__ctk_fault__" in w) {
      // NOW-10 fault injection: exercise §6.3 interceptor_failed.
      throw new Error("ctk scripted fault: raise");
    }
    return w;
  }
}

/** Wraps the scripted interceptor and records every ctx passed. */
class RecordingInterceptor extends ScriptedInterceptor {
  readonly recorded: AgentContext[] = [];
  override intercept(ctx: AgentContext): Verdict {
    this.recorded.push(JSON.parse(JSON.stringify(ctx)));
    return super.intercept(ctx);
  }
}

class ScriptedResolver {
  private readonly rulesJson: string;
  constructor(rules: JsonValue) {
    this.rulesJson = JSON.stringify(rules);
  }
  resolve(req: ApprovalRequest): ApprovalResolution {
    // §10.1: identity may be null (null provider). The scripted engine
    // works in strings; "" round-trips to null below.
    const requestIdentity = req.context_identity ?? "";
    const r = JSON.parse(
      native.ctkScriptedResolve(this.rulesJson, JSON.stringify(req.context), requestIdentity),
    );
    if (r !== null && typeof r === "object" && "__ctk_fault__" in r) {
      // NOW-10 fault injection: exercise §9 approval_resolver_failed.
      throw new Error("ctk scripted fault: raise");
    }
    if (r.context_identity === "" && req.context_identity === null) {
      r.context_identity = null;
    }
    return r;
  }
}

function runRecordToWire(rr: RunRecord): string {
  return JSON.stringify({
    outcome: rr.outcome,
    final_output: rr.final_output ?? null,
    tool_invocations: rr.tool_invocations,
    error: rr.error ?? null,
    identities: rr.identities.map(([i, e]) => ({ input_identity: i, enforced_identity: e })),
  });
}

export async function runVector(harness: Harness, vector: JsonValue): Promise<VectorResult> {
  const v = vector as Record<string, JsonValue>;
  const vectorJson = JSON.stringify(vector);

  if (TODO_STAGE_4.some((id) => (v.id as string).startsWith(id))) {
    return {
      id: v.id as string,
      title: v.title as string,
      status: "skip",
      detail: "TODO(stage-4): stale pre-P-003 vector, rewritten in stage 4",
      failures: [],
    };
  }

  const capsJson = JSON.stringify([...harness.capabilities].sort());
  const skip = JSON.parse(native.ctkShouldSkip(vectorJson, capsJson));
  if (skip !== null) {
    return {
      id: v.id as string,
      title: v.title as string,
      status: "skip",
      detail: skip,
      failures: [],
    };
  }

  // Multi-interceptor vectors (§7.1 fold-through) use interceptor_scripts;
  // single-interceptor vectors use interceptor_script. Only the FIRST
  // interceptor records: expect.interceptions describes each emission as
  // the first-registered interceptor saw it. An empty interceptor_scripts
  // registers zero interceptors (§7 fail-closed vector).
  const scripts = (v.interceptor_scripts as JsonValue[] | undefined) ?? [v.interceptor_script];
  const first = scripts.length > 0 ? new RecordingInterceptor(scripts[0]) : null;
  const interceptors: Interceptor[] = first ? [first] : [];
  for (const s of scripts.slice(1)) interceptors.push(new ScriptedInterceptor(s));

  const approval = v.approval_script as JsonValue[] | undefined;
  const resolver = approval ? new ScriptedResolver(approval) : null;
  const mode = ((v.mode as string) ?? "enforce") as EnforcementMode;
  // §13.2: composition vectors carry the profile/knobs they apply to;
  // absent means the pre-P-003 default (`sequential/first_deny, stop`).
  const composition =
    (v.composition as unknown as CompositionConfig | undefined) ?? Composition.default();

  harness.setup(v.scenario as unknown as Scenario, interceptors, resolver, mode, composition);
  let rr: RunRecord;
  try {
    rr = await harness.run();
  } catch (e) {
    return {
      id: v.id as string,
      title: v.title as string,
      status: "fail",
      detail: "",
      failures: [`harness.run threw: ${e}`],
    };
  } finally {
    harness.teardown();
  }

  return JSON.parse(
    native.ctkAssert(vectorJson, JSON.stringify(first?.recorded ?? []), runRecordToWire(rr)),
  );
}

export async function runVectors(
  harnessFactory: () => Harness,
  vectors: JsonValue[],
): Promise<VectorResult[]> {
  const out: VectorResult[] = [];
  for (const v of vectors) {
    out.push(await runVector(harnessFactory(), v));
  }
  return out;
}

File: sdk/dotnet/src/AgentHooks.Conformance/Runner.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
// CTK runner: load vectors, drive an IHarness, assert `expect`.
//
// The assertion engine, capability skip check, and scripted
// interceptor/resolver evaluation live in the Rust core (Native.Ctk*).
// This class keeps only vector globbing, the harness setup/run/teardown
// orchestration (native callbacks), and RunRecord marshalling.
// See conformance/RUNNER.md.

using System.Text.Json;
using System.Text.Json.Nodes;

namespace AgentHooks.Conformance;

public sealed record VectorResult(
    string Id, string Title, string Status,
    string Detail, IReadOnlyList<string> Failures);

public static class Runner
{
    private static readonly JsonSerializerOptions Compact = new() { WriteIndented = false };

    public static IEnumerable<JsonObject> LoadVectors(string directory)
    {
        foreach (var f in Directory.EnumerateFiles(directory, "AH-CTK-*.json").OrderBy(p => p))
            yield return (JsonObject)JsonNode.Parse(File.ReadAllText(f))!;
    }

    public static async Task<VectorResult> RunVectorAsync(
        IHarness harness, JsonObject vector, CancellationToken ct = default)
    {
        var id = (string)vector["id"]!;
        var title = (string)vector["title"]!;
        var vectorJson = vector.ToJsonString(Compact);

        // Capability skip via core.
        var caps = new JsonArray(
            harness.Capabilities.Select(c => (JsonNode)c.ToWireName()).ToArray());
        var skip = JsonNode.Parse(
            Native.CtkShouldSkip(vectorJson, caps.ToJsonString(Compact)));
        if (skip is JsonValue sv && sv.TryGetValue(out string? reason))
            return new VectorResult(id, title, "skip", reason!, []);

        var scenario = Scenario.FromWire((JsonObject)vector["scenario"]!);

        // Multi-interceptor vectors (§7.1 fold-through) use
        // interceptor_scripts; single-interceptor vectors use
        // interceptor_script. Only the FIRST interceptor records:
        // expect.interceptions describes each emission as the
        // first-registered interceptor saw it. An empty
        // interceptor_scripts registers zero interceptors (§7
        // fail-closed vector).
        List<JsonArray> scripts;
        if (vector["interceptor_scripts"] is JsonArray multi)
            scripts = multi.Select(s => (JsonArray)s!.DeepClone()).ToList();
        else
            scripts = [(JsonArray)(vector["interceptor_script"] ?? new JsonArray()).DeepClone()];

        var recorded = new List<JsonObject>();
        var interceptors = new List<IInterceptor>();
        if (scripts.Count > 0)
        {
            interceptors.Add(new RecordingScriptedInterceptor(
                scripts[0].ToJsonString(Compact), recorded));
            foreach (var s in scripts.Skip(1))
                interceptors.Add(new ScriptedInterceptor(s.ToJsonString(Compact)));
        }

        var approval = vector["approval_script"] as JsonArray;
        IApprovalResolver? resolver = approval is { Count: > 0 }
            ? new ScriptedResolver(approval.ToJsonString(Compact))
            : null;

        var mode = (string?)vector["mode"] == "evaluate_only"
            ? EnforcementMode.EvaluateOnly : EnforcementMode.Enforce;

        harness.Setup(scenario, interceptors, resolver, mode);
        RunRecord rr;
        try
        {
            rr = await harness.RunAsync(ct);
        }
        catch (Exception e)
        {
            return new VectorResult(id, title, "fail", "",
                [$"harness.RunAsync raised: {e}"]);
        }
        finally
        {
            harness.Teardown();
        }

        var recordedJson = new JsonArray(
            recorded.Select(c => (JsonNode)c).ToArray()).ToJsonString(Compact);
        var rrJson = RunRecordToWire(rr);
        var result = (JsonObject)JsonNode.Parse(
            Native.CtkAssert(vectorJson, recordedJson, rrJson))!;
        return new VectorResult(
            (string)result["id"]!,
            (string)result["title"]!,
            (string)result["status"]!,
            (string?)result["detail"] ?? "",
            (result["failures"] as JsonArray)?.Select(n => (string)n!).ToList() ?? []);
    }

    private static string RunRecordToWire(RunRecord rr)
    {
        var identities = new JsonArray();
        foreach (var (i, e) in rr.Identities ?? [])
            identities.Add(new JsonObject
            {
                ["input_identity"] = i, ["enforced_identity"] = e,
            });
        var o = new JsonObject
        {
            ["outcome"] = rr.Outcome.ToWireName(),
            ["final_output"] = rr.FinalOutput?.DeepClone(),
            ["tool_invocations"] = new JsonArray(
                rr.ToolInvocations.Select(t => (JsonNode)t.DeepClone()).ToArray()),
            ["error"] = rr.Error,
            ["identities"] = identities,
        };
        return o.ToJsonString(Compact);
    }

    /// <summary>Replays one interceptor rule list via the Rust core.</summary>
    private class ScriptedInterceptor(string rulesJson) : IInterceptor
    {
        protected readonly string RulesJson = rulesJson;

        public virtual ValueTask<Verdict> InterceptAsync(
            AgentContext ctx, CancellationToken ct = default)
        {
            var w = (JsonObject)JsonNode.Parse(
                Native.CtkScriptedIntercept(RulesJson, ctx.Json.ToJsonString(Compact)))!;
            if (w.ContainsKey("__ctk_fault__"))
            {
                // NOW-10 fault injection: exercise §6.3 interceptor_failed.
                throw new InvalidOperationException("ctk scripted fault: raise");
            }
            return ValueTask.FromResult(Verdict.FromWire(w));
        }
    }

    /// <summary>Records every ctx (deep copy) then replays the rules.</summary>
    private sealed class RecordingScriptedInterceptor(
        string rulesJson, List<JsonObject> recorded) : ScriptedInterceptor(rulesJson)
    {
        public override ValueTask<Verdict> InterceptAsync(
            AgentContext ctx, CancellationToken ct = default)
        {
            recorded.Add((JsonObject)ctx.Json.DeepClone());
            return base.InterceptAsync(ctx, ct);
        }
    }

    /// <summary>Replays the vector's approval_script via the Rust core.</summary>
    private sealed class ScriptedResolver(string rulesJson) : IApprovalResolver
    {
        public ValueTask<ApprovalResolution> ResolveAsync(
            ApprovalRequest req, CancellationToken ct = default)
        {
            var r = (JsonObject)JsonNode.Parse(
                Native.CtkScriptedResolve(
                    rulesJson, req.Context.Json.ToJsonString(Compact), req.ContextIdentity))!;
            if (r.ContainsKey("__ctk_fault__"))
            {
                // NOW-10 fault injection: exercise §9 approval_resolver_failed.
                throw new InvalidOperationException("ctk scripted fault: raise");
            }
            var outcome = (string)r["outcome"]! switch
            {
                "approve" => ApprovalOutcome.Approve,
                "reject" => ApprovalOutcome.Reject,
                _ => ApprovalOutcome.Unresolved,
            };
            var v = r["verdict"] is JsonObject vw ? Verdict.FromWire(vw) : null;
            return ValueTask.FromResult(
                new ApprovalResolution(outcome, (string)r["context_identity"]!, v));
        }
    }
}
